x-pack/plugins/fleet/server/services/agents/status.ts

```diff
@@ -37,88 +41,108 @@ export async function getAgentStatusById(
-}
-
 export async function getAgentStatusForAgentPolicy(
   esClient: ElasticsearchClient,
+  soClient: SavedObjectsClientContract,
   agentPolicyId?: string,
   filterKuery?: string
 ) {
-  let pitId: string | undefined;
+  const logger = appContextService.getLogger();
+  const runtimeFields = await buildAgentStatusRuntimeField(soClient);
+
+  const clauses: QueryDslQueryContainer[] = [];
+
+  if (filterKuery) {
+    const kueryAsElasticsearchQuery = toElasticsearchQuery(
+      fromKueryExpression(removeSOAttributes(filterKuery))
+    );
+    clauses.push(kueryAsElasticsearchQuery);
+  }
+
+  if (agentPolicyId) {
+    clauses.push({
+      term: {
+        policy_id: agentPolicyId,
+      },
+    });
+  }
+
+  const query =
+    clauses.length > 0
+      ? {
+          bool: {
+            must: clauses,
+          },
+        }
+      : undefined;
+
+  const statuses: Record<AgentStatus, number> = {
+    online: 0,
+    error: 0,
+    inactive: 0,
+    offline: 0,
+    updating: 0,
+    unenrolled: 0,
+    degraded: 0,
+    enrolling: 0,
+    unenrolling: 0,
+  };
+
+  let response;
+
   try {
-    pitId = await openPointInTime(esClient);
+    response = await esClient.search<
+      null,
+      { status: AggregationsTermsAggregateBase<AggregationsStatusTermsBucketKeys> }
+    >({
+      index: AGENTS_INDEX,
+      size: 0,
+      query,
+      fields: Object.keys(runtimeFields),
+      runtime_mappings: runtimeFields,
+      aggregations: {
+        status: {
+          terms: {
+            field: 'status',
+            size: Object.keys(statuses).length,
+          },
+        },
+      },
+    });
   } catch (error) {
-    if (error.statusCode === 404) {
-      appContextService
-        .getLogger()
-        .debug('Index .fleet-agents does not exist yet, skipping point in time.');
-    } else {
-      throw error;
-    }
+    logger.error(`Error getting agent statuses: ${error}`);
+    throw error;
   }
-  const [all, allActive, online, error, offline, updating] = await pMap(
-    [
-      undefined, // All agents, including inactive
-      undefined, // All active agents
-      AgentStatusKueryHelper.buildKueryForOnlineAgents(),
-      AgentStatusKueryHelper.buildKueryForErrorAgents(),
-      AgentStatusKueryHelper.buildKueryForOfflineAgents(),
-      AgentStatusKueryHelper.buildKueryForUpdatingAgents(),
-    ],
-    (kuery, index) =>
-      getAgentsByKuery(esClient, {
-        showInactive: index === 0,
-        perPage: 0,
-        page: 1,
-        pitId,
-        kuery: joinKuerys(
-          ...[
-            kuery,
-            filterKuery,
-            agentPolicyId ? `${AGENTS_PREFIX}.policy_id:"${agentPolicyId}"` : undefined,
-          ]
-        ),
-      }),
-    {
-      concurrency: 1,
-    }
-  );
 
-  if (pitId) {
-    await closePointInTime(esClient, pitId);
+  const buckets = response?.aggregations?.status?.buckets || [];
+
+  if (!Array.isArray(buckets)) {
+    throw new Error('buckets is not an array');
```

Comment:
I'm not sure if this can ever happen.

Reply:
You're right, this was a temporary hack to get around a TS error 😅